Fix multi pages search for gh search

leudz opened the pull request:
This fixes how `gh search` retrieves items when more than 100 items are requested and not a multiple of 100.

When using pagination, two variables are used: page and perPage. Pages can contain a maximum of 100 items.
The previous implementation tried to save the API some work by requesting less items for the last page.
But it didn't handle pagination correctly and ended up requesting some items a second time.

For example if 150 items are queried, the first page returns items 0->99.
Then it would try to query 50 items by asking for page 2 with 50 per page.
But this would return items 50->99 which isn't what we want.

This new version instead picks a fixed amount per page, in this case 100.
The first page returns items 0->99 and the second 100->199.
Then we discard items 150->199.

When less than 100 items are requested there is no change, we only request what we need.

----

In theory it's possible to both save the API some work and keep track of pagination but that would be more complex.

Fixes #9749

leudz committed:
Fix multi pages search for gh search

leudz commented on pkg/search/searcher.go:
```diff
@@ -60,96 +61,138 @@ func NewSearcher(client *http.Client, host string) Searcher {
+		// otherwise add all the results.
+		itemsToAdd := min(len(page.Items), toRetrieve)
+
+		result.Total += itemsToAdd
```
I'm not 100% sure what this total is.
I assume it's the total number of elements returned.

It used to only return the last page's total which affect the pagination test.

andyfeller replied:
`Total` is used in displaying the results for communicating how many results out of the total results are being shown.

For example, `gh search code`:

https://github.com/cli/cli/blob/072534c388cf20924be3e359819c2674d6d4190a/pkg/cmd/search/code/code.go#L136-L163

andyfeller replied:
**issue:** I have reverted this logic because it causes the `X of Y` results header to report `--limit` for Y instead of the total number of matching search results.

```
		result.Total = page.Total
```

### Demo

Below, we see this 150-limited code search reporting exactly 150 total matches are found:

```shell
$ bin/gh search code "fmt.Fprintf" --repo cli/cli --limit 150

Showing 150 of 150 results

cli/cli script/build.go
	fmt.Fprintf(os.Stderr, "Don't know how to build task `%s`.\n", task)
	fmt.Fprintf(os.Stderr, "%s: building task `%s` failed.\n", self, task)
```

with original logic, we see `150 of 161 results` which is the total number of matching search entries:

```shell
$ bin/gh search code "fmt.Fprintf" --repo cli/cli --limit 150

Showing 150 of 161 results

cli/cli script/build.go
	fmt.Fprintf(os.Stderr, "Don't know how to build task `%s`.\n", task)
	fmt.Fprintf(os.Stderr, "%s: building task `%s` failed.\n", self, task)
```

leudz replied:
Okay so it's the total matching the search query... sorry for the useless changes. Would it be worth add a small comment on the field?

andyfeller replied:
I think that is a solid call out as it really isn't obvious, yeah 💯

leudz commented on pkg/search/searcher_test.go:
```diff
@@ -87,20 +103,64 @@ func TestSearcherCode(t *testing.T) {
+				reg.Register(firstReq, firstRes)
+				reg.Register(secondReq, secondRes)
+			},
+		},
```
There are only 4 small changes in this tests.
`per_page` is 30, both in the header and the second request.

The total for the first and second responses have gone from 2 to 1.
This link to a previous review comment. I assume this total is the number of items in the response.
In that case, it has always been 1.

andyfeller replied:
@leudz : As the `.Total` logic changes needed to be reverted, there are going to be some changes to these tests as `.Total` is not the number of items in the page, but the total number of matches to the search query.

Additionally, I've made some enhancements around test table data initialization.

andyfeller commented:
Thank you for your patience and efforts to improve the GitHub CLI, @leudz! ❤️  I'm working to provide some initial feedback today so we can get this shaped up and shipped out. 💪

andyfeller committed:
Restore result.Total logic, fix formatting

This change restores the original logic of passing the search total count logic as is to the result.

Additionally, this undoes some of the contributor's formatting changes that increase the changed lines to review.

andyfeller committed:
Fix remaining search logic totals

andyfeller commented:
## Demo

After restoring the `.Total` logic, I wanted to manually confirm the experience relative to the GitHub REST Search API results.

### `gh search code`

```shell
$ gh api "search/code?q=fmt.Fprintf+repo:cli/cli" -X GET --jq '.total_count'
161

$ gh search code fmt.Fprintf --repo cli/cli

Showing 30 of 161 results
```

### `gh search commits`

```shell
$ gh api "search/commits?q=author:andyfeller+repo:cli/cli" -X GET --jq '.total_count'
335

$ gh search commits --repo cli/cli author:andyfeller      

Showing 30 of 335 commits
```

### `gh search issues`

```shell
$ gh api "search/issues?q=author:andyfeller+repo:cli/cli+type:issue" -X GET --jq '.total_count'
39

$ gh search issues --repo cli/cli author:andyfeller                                            

Showing 30 of 39 issues
```

### `gh search prs`

```shell
$ gh api "search/issues?q=author:andyfeller+repo:cli/cli+type:pr" -X GET --jq '.total_count'
51

$ gh search prs --repo cli/cli author:andyfeller                                       

Showing 30 of 51 pull requests
```

### `gh search repos`

```shell
$ gh api "search/repositories?q=user:andyfeller" -X GET --jq '.total_count'
65

$ gh search repos --owner andyfeller

Showing 30 of 65 repositories
```

andyfeller committed:
Fix search tests around totals, initialization

This commit is focused on fixing the `searcher` tests for a few reasons:

1. Correcting the `.Total` logic in the previous commit caused changes to tests to fail
2. Tests involving results that exceed the max per page have been improved with new initialize helper, allowing testing table scenarios to be self contained
3. Tests that stub JSON response payloads have been standardized on maps rather than GitHub type primitives (Repository, Issue, Commit, Code, etc)
4. Tests had some minor formatting changes to make them easier to understand and maintain

andyfeller committed:
Update searcher tests

This commit moves the remaining searcher tests from using JSON marshaled types to using JSON responses for consistency.  There appears to be a weird JSON marshaling error with search.Repository that does not map `Name` field in the process.

Additionally, the test scenarios around pulling multiple pages beneath the total results have been updated to demonstrate that the REST API returns full pages in both of these cases, which is below the total number of results.

andyfeller commented on pkg/search/searcher_test.go:
```diff
@@ -435,13 +575,73 @@ func TestSearcherRepositories(t *testing.T) {
+				Total: 287,
+			},
+			httpStubs: func(reg *httpmock.Registry) {
+				firstReq := httpmock.QueryMatcher("GET", "search/repositories", url.Values{
+					"page":     []string{"1"},
+					"per_page": []string{"100"},
+					"order":    []string{"desc"},
+					"sort":     []string{"stars"},
+					"q":        []string{"keyword stars:>=5 topic:topic"},
+				})
+				firstRes := httpmock.JSONResponse(map[string]interface{}{
+					"incomplete_results": false,
+					"total_count":        287,
+					"items": initialize(0, 100, func(i int) interface{} {
+						return map[string]interface{}{
+							"name": fmt.Sprintf("name%d", i),
+						}
+					}),
+				})
+				firstRes = httpmock.WithHeader(firstRes, "Link", `<https://api.github.com/search/repositories?page=2&per_page=100&q=org%3Agithub>; rel="next"`)
+				secondReq := httpmock.QueryMatcher("GET", "search/repositories", url.Values{
+					"page":     []string{"2"},
+					"per_page": []string{"100"},
+					"order":    []string{"desc"},
+					"sort":     []string{"stars"},
+					"q":        []string{"keyword stars:>=5 topic:topic"},
+				})
+				secondRes := httpmock.JSONResponse(map[string]interface{}{
+					"incomplete_results": false,
+					"total_count":        287,
+					"items": initialize(100, 200, func(i int) interface{} {
+						return map[string]interface{}{
+							"name": fmt.Sprintf("name%d", i),
+						}
+					}),
+				})
+				reg.Register(firstReq, firstRes)
+				reg.Register(secondReq, secondRes)
+			},
+		},
```
**concern:** originally, these tests relied upon the `httpmock.Registry` to JSON marshal the `*Result` types to JSON, however there is something about the `search.Repository` type where `Name` field was not being preserved in the marshaling.

Rather than having inconsistent test setups, I aligned the tests on capturing the JSON we expect from the REST API.

andyfeller replied:
Closing the loop here, it appears that `search.Repository` JSON marshaling is a problem because of this logic used for exporting `gh search code` as it also contains a repository field:

https://github.com/cli/cli/blob/5d8dbc0d534d71d1b011e52016fe17b6565a5317/pkg/search/result.go#L356-L385

Given this surprising behavior, the move away from using the `Results` type for response mocking makes more sense.

andyfeller commented on pkg/search/searcher_test.go:
```diff
@@ -435,13 +575,73 @@ func TestSearcherRepositories(t *testing.T) {
+				secondRes := httpmock.JSONResponse(map[string]interface{}{
+					"incomplete_results": false,
+					"total_count":        287,
+					"items": initialize(100, 200, func(i int) interface{} {
```
**thought:** while discussing these tests with @jtmcg, Tyler made a call out how the REST API will return a full second page.  He suggested initializing this to be a full page, which is closer to the behavior we will encounter in use.

andyfeller commented on pkg/search/searcher_test.go:
```diff
@@ -435,13 +575,73 @@ func TestSearcherRepositories(t *testing.T) {
 			},
 		},
+		{
+			name: "collect full and partial pages under total number of matching search results",
```
**thought:** after talking with @jtmcg, Tyler suggested renaming these test scenarios to explain they are ensuring these commands process REST responses and return a collection of full and partial pages to the user while still being beneath the total number of matching search results.

andyfeller commented on pkg/search/searcher_test.go:
```diff
@@ -239,27 +315,92 @@ func TestSearcherCommits(t *testing.T) {
+				})
+				firstRes := httpmock.JSONResponse(map[string]interface{}{
+					"incomplete_results": false,
+					"total_count":        287,
```
**thought:** the REST search API returns the total number of matching search results on every page.  In order to have a number stand out separate from the client-side limit, Tyler suggested 287, which is as good as any.

leudz replied:
That's a very nice change 👍

andyfeller commented:
@leudz : Just because I did a bit more changes to this PR rather than opening up a ton of suggestions, I've asked a fellow maintainer to be the final PR reviewer as I'd value a 2nd set of eyes. 🙇 Thank you again for your patience!

andyfeller committed:
PR feedback

- update local variables to communicate what they are
- added docblock explaining search results populated

leudz committed:
Add comments for Total

leudz commented:
@andyfeller I've added a few comments for Total like we talked about.
With the doc on `search` you've added it's a bit less valuable but it might still avoid someone from missing it.

BagToad reviewed (approved):
LGTM! ✨ The problem and fix make sense, and this branch does fix the problem when I tested it out myself.
